tests/test_directives.py

```diff
@@ -0,0 +1,266 @@
+    def test_get_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        for item in available_directives_list:
+            assert directives[item.name] == item.default
+
+    def test_add_directive(self, available_directives_list):
+        directives = Directives(available_directives_list[:-1])
+        directives._add_directive(PROCESS_FRACTION)
+        assert directives[PROCESS_FRACTION.name] == PROCESS_FRACTION.default
+        with pytest.raises(TypeError):
+            directives._add_directive('Test')
+        with pytest.raises(ValueError):
+            directives._add_directive(PROCESS_FRACTION)
+
+    def test_set_defined_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives._set_defined_directive(NP.name, 10)
+        assert directives[NP.name] == 10
+
+    def test_set_defined_directive_invalid(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        with pytest.raises(ValueError):
+            directives._set_defined_directive(NP.name, 0)
+
+    def test_set_undefined_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        with pytest.raises(Exception):
+            directives._set_defined_directive('test', 0)
+
+    def test_set_directives_item(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives['test'] = True
+        assert directives['test']
+
+    def test_del_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives['test'] = True
+        directives._set_defined_directive(NP.name, 100)
+        assert directives[NP.name] == 100
+        assert directives['test']
+        del directives[NP.name]
+        assert directives[NP.name] == NP.default
+        del directives['test']
+        with pytest.raises(KeyError):
+            directives['test']
+
+    def test_update_directive_without_aggregate(
+        self, available_directives_list, non_default_directive_values
+    ):
+        directives = Directives(available_directives_list)
+        valid_values_0 = non_default_directive_values()
+        valid_values_1 = non_default_directive_values(1)
+        directives.update(valid_values_0)
+        expected_values = {'np': 4, 'ngpu': 1, 'nranks': 0,
+                           'omp_num_threads': 10, 'executable': 'PathFinder',
+                           'walltime': 20.0, 'memory': 16, 'processor_fraction': 0.5}
+        directives.update(valid_values_1)
+        for dirs in directives:
+            assert directives[dirs] == expected_values[dirs]
+
+    def test_update_directive_serial(self, available_directives_list, non_default_directive_values):
+        directives1 = Directives(available_directives_list)
+        directives2 = Directives(available_directives_list)
```

Comment:
Having seen how often this pattern arises, I have changed my mind. I like the idea of a `directives` fixture so we don't have to pass `available_directives_list` everywhere.